File: 백준/Gold/2023. 신기한 소수/신기한 소수.java
import java.util.Scanner;

public class Main {
    static int n;

    static void dfs(int num, int digit) {
        if (digit == n && isPrime(num)) {
            System.out.println(num);
            return;
        }

        for (int i = 1; i < 10; i++) {
            if (i % 2 == 0) continue;

            if (isPrime(num * 10 + i)) {
                dfs(num * 10 + i, digit + 1);
            }
        }
    }

    static boolean isPrime(int num) {
        for (int i = 2; i <= num / 2; i++) {
            if (num % i == 0) return false;
        }

        return true;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();

        dfs(2, 1);
        dfs(3, 1);
        dfs(5, 1);
        dfs(7, 1);
    }
}

File: 백준/Gold/2023. 신기한 소수/신기한 소수.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.math.sqrt

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    fun isPrime(num: Int): Boolean {
        if (num < 2) return false
        if (num == 2) return true
        if (num % 2 == 0) return false

        for (i in 3..sqrt(num.toDouble()).toInt() step 2) {
            if (num % i == 0) return false
        }

        return true
    }

    fun dfs(num: Int, digit: Int) {
        if (digit == n) {
            println(num)
            return
        }

        for (i in 1..9 step 2) {
            val nextNum = num * 10 + i
            if (isPrime(nextNum)) dfs(nextNum, digit + 1)
        }
    }

    listOf(2, 3, 5, 7).forEach { dfs(it, 1) }
}